File: scripts/cross_target.mjs
// Cross-target reproducibility runner (WebAssembly via Node).
//
// Loads the datacube-rs WASM module (the same one that powers the browser
// demo) and runs the five per-series estimators over the shared input. The
// WASM, native and Python targets share one Rust core, so results agree to
// machine precision. Prints results + per-estimator timing as JSON.
//
// Build:  wasm-pack build --target nodejs --out-dir /tmp/wasm-node  (from crates/datacube-wasm)
// Run:    node scripts/cross_target.mjs /tmp/wasm-node /tmp/xt_input.json

import { readFileSync } from "node:fs";
import { performance } from "node:perf_hooks";

const pkgDir = process.argv[2] || "/tmp/wasm-node";
const inputPath = process.argv[3] || "/tmp/xt_input.json";
const wasm = await import(`${pkgDir}/datacube_wasm.js`);

const inp = JSON.parse(readFileSync(inputPath, "utf8"));
const nan = (a) => Float64Array.from(a.map((v) => (v === null ? NaN : v)));
const t = nan(inp.t), y = nan(inp.y);

const lt = wasm.linear_trend(t, y);
const ts = wasm.theil_sen(t, y);
const mk = wasm.mann_kendall(y, 0.05);
const hr = wasm.harmonic_regression(t, y, 1.0, 2);
const br = wasm.detect_breaks(t, y, 0.05, 1, 1.0, 12);

const reps = 2000;
const us = (f) => {
  const s = performance.now();
  for (let i = 0; i < reps; i++) f();
  return (performance.now() - s) * 1e3 / reps; // µs/call
};

const out = {
  target: "wasm",
  results: {
    "linear_trend.slope": lt.slope,
    "linear_trend.p_value": lt.p_value,
    "theil_sen.slope": ts.slope,
    "mann_kendall.tau": mk.tau,
    "mann_kendall.z": mk.z,
    "mann_kendall.p_value": mk.p_value,
    "harmonic.slope": hr.slope,
    "harmonic.amplitude_1": hr.components[0].amplitude,
    "harmonic.rmse": hr.rmse,
    "detect_breaks.statistic": br.statistic,
    "detect_breaks.p_value": br.p_value,
    "detect_breaks.n_breaks": br.breaks.length,
  },
  timing: {
    linear_trend_us: us(() => wasm.linear_trend(t, y)),
    theil_sen_us: us(() => wasm.theil_sen(t, y)),
    mann_kendall_us: us(() => wasm.mann_kendall(y, 0.05)),
    harmonic_us: us(() => wasm.harmonic_regression(t, y, 1.0, 2)),
    detect_breaks_us: us(() => wasm.detect_breaks(t, y, 0.05, 1, 1.0, 12)),
  },
};
console.log(JSON.stringify(out, null, 2));

File: scripts/cross_target.py
#!/usr/bin/env python3
"""Cross-target reproducibility runner (Python / PyO3).

Reads the shared input and runs the five per-series estimators through the
datacube_rs Python extension. Companion to the native example and the Node
WebAssembly runner; all three share one Rust core, so the results are
bit-identical. Prints results + per-estimator timing as JSON.

Run: .venv-validate/bin/python scripts/cross_target.py /tmp/xt_input.json
"""
import json
import sys
import time

import numpy as np
import datacube_rs as dc

inp = json.load(open(sys.argv[1] if len(sys.argv) > 1 else "/tmp/xt_input.json"))
t = np.array([np.nan if v is None else v for v in inp["t"]], dtype=np.float64)
y = np.array([np.nan if v is None else v for v in inp["y"]], dtype=np.float64)

lt = dc.linear_trend(t, y)
ts = dc.theil_sen(t, y)
mk = dc.mann_kendall(y, 0.05)
hr = dc.harmonic_regression(t, y, 1.0, 2)
br = dc.detect_breaks(t, y, 0.05, 1, 1.0, 12)

reps = 2000
def us(f):
    s = time.perf_counter()
    for _ in range(reps):
        f()
    return (time.perf_counter() - s) * 1e6 / reps

out = {
    "target": "python",
    "results": {
        "linear_trend.slope": lt["slope"],
        "linear_trend.p_value": lt["p_value"],
        "theil_sen.slope": ts["slope"],
        "mann_kendall.tau": mk["tau"],
        "mann_kendall.z": mk["z"],
        "mann_kendall.p_value": mk["p_value"],
        "harmonic.slope": hr["slope"],
        "harmonic.amplitude_1": hr["components"][0]["amplitude"],
        "harmonic.rmse": hr["rmse"],
        "detect_breaks.statistic": br["statistic"],
        "detect_breaks.p_value": br["p_value"],
        "detect_breaks.n_breaks": len(br["breaks"]),
    },
    "timing": {
        "linear_trend_us": us(lambda: dc.linear_trend(t, y)),
        "theil_sen_us": us(lambda: dc.theil_sen(t, y)),
        "mann_kendall_us": us(lambda: dc.mann_kendall(y, 0.05)),
        "harmonic_us": us(lambda: dc.harmonic_regression(t, y, 1.0, 2)),
        "detect_breaks_us": us(lambda: dc.detect_breaks(t, y, 0.05, 1, 1.0, 12)),
    },
}
print(json.dumps(out, indent=2, sort_keys=True))
